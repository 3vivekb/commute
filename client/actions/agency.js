import fetchAction from '@conveyal/woonerf/fetch'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addLocally = createAction('add agency')
export const createAgency = (newAgency) => {
  newAgency.id = uuid.v4()
  newAgency.organizations = []
  return [
    addLocally(newAgency),
    push(`/agency/${newAgency.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete agency')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/agency/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteAgency = (id) => [deleteLocally(id), push('/')]

export const loadAgencies = () => fetchAction({
  url: '/api/agency',
  next: (res, err) => {
    if (!err) {
      return setAllLocally(res)
    }
  }
})

export const updateAgency = (agency) => [
  setLocally(agency),
  push(`/agency/${agency.id}`)
]
/* const updateOnServer = (agency) =>
  serverAction({
    url: `/api/agency/${agency.id}`,
    options: {
      body: JSON.stringify(agency),
      method: 'put'
    }
  }) */  // TODO: update on server

export const setLocally = createAction('set agency')
export const setAllLocally = createAction('set agencies')
