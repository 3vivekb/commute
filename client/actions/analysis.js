import moment from 'moment'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addAnalysis = createAction('add analysis')
export const createAnalysis = (newAnalysis, organizationId) => {
  newAnalysis.id = uuid.v4()
  newAnalysis.lastRunDateTime = moment().unix()
  newAnalysis.trips = []
  return [
    addAnalysis({ analysis: newAnalysis, organizationId }),
    push(`/organizations/${organizationId}/analysis/${newAnalysis.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete analysis')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/analysis/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteAnalysis = (id, organzationId) => [
  deleteLocally({ analysisId: id, organzationId }),
  push(`/organizations/${organzationId}`)
]
