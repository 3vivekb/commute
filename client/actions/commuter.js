import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addCommuter = createAction('add commuter')
export const createCommuter = (commuter, groupId) => {
  commuter.groupId = groupId
  commuter.id = uuid.v4()
  return [
    addCommuter(commuter),
    push(`/group/${groupId}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete commuter')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/commuter/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteCommuter = (commuterId, groupId) => [
  deleteLocally(commuterId),
  push(`/group/${groupId}`)
]

const updateLocally = createAction('update commuter')
/* const updateOnServer = (id) =>
  serverAction({
    url: `/api/commuter/${id}`,
    options: {
      method: 'update'
    }
  }) */ // TODO update on server
export const updateCommuter = (commuter, groupId) => [
  updateLocally(commuter),
  push(`/group/${groupId}`)
]
