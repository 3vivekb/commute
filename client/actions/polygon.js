import makeGenericModelActions from '../utils/actions'

const actions = makeGenericModelActions({
  commands: {
    'Collection DELETE': {},
    'Collection GET': {}
  },
  pluralName: 'polygons',
  singularName: 'polygon'
})

export default actions
