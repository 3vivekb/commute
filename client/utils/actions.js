import fetchAction from '@conveyal/woonerf/fetch'
import qs from 'qs'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'

/**
 * Make generic model actions that communicate with a rest server
 *
 * @param  {Object} cfg An object with the following parameters:
 * - commands: An object with the keys being commands and the corresponding cfg of each command
 * -- redirectionStrategy: The redirection strategy to use upon success.
 *  Can be one of following: 'toEntity', 'toParent', 'toHome', undefined (no redirection)
 *  Defaults are as follows:
 *    Collection GET, GET: null
 *    Collection POST, PUT: toEntity
 *    DELETE: toParent
 * - parentKey: Key in model of parent's id
 *   (must be provided if using toParent redirection strategy)
 * - parentName: Name of app route to parent view
 *   (must be provided if using toParent redirection strategy)
 * - pluralName: Plural name of the model
 * - singularName: Singular name of the model
 * @return {Object}     The actions
 */
const makeGenericModelActions = (cfg) => {
  const {commands, parentKey, parentName, pluralName, singularName} = cfg
  const actions = {}

  // make local set actions
  const addLocally = createAction(`add ${singularName}`)
  const deleteLocally = createAction(`delete ${singularName}`)
  const setLocally = createAction(`set ${singularName}`)
  const setAllLocally = createAction(`set ${pluralName}`)

  const baseEndpoint = `/api/${singularName}`

  const redirectionStrategies = {
    'toEntity': (entity) => {
      return push(`/${singularName}/${entity._id}`)
    },
    'toHome': () => {
      return push('/')
    },
    'toParent': (entity) => {
      return push(`/${parentName}/${entity[parentKey]}`)
    }
  }

  /**
   * Do a redirect if needed
   *
   * @param  {Object} cfg Paremters as follows:
   *   - actions: The actions to append to
   *   - endpointCfg: The endpoint config
   *   - defaultStrategy: The default redire strategy to use (default: undefined)
   *   - redirectArgs: Arguments to provide to the redirectionStrategy
   */
  const doRedirectIfNecessary = (cfg) => {
    const redirectionStrategy = (cfg.endpointCfg.redirectionStrategy !== undefined
      ? cfg.endpointCfg.redirectionStrategy
      : cfg.defaultStrategy)

    if (redirectionStrategy) {
      const redirect = redirectionStrategies[redirectionStrategy](cfg.redirectArgs)
      if (redirect) {
        cfg.actions.push(redirect)
      }
    }
  }

  if (commands['Collection GET']) {
    actions.loadAll = (queryParams) => {
      // only include filteredKeys in querystring
      let queryString = ''
      if (queryParams) {
        queryString = qs.stringify(queryParams)
        if (queryString.length !== 0) {
          queryString = '?' + queryString
        }
      }

      // make request
      return fetchAction({
        next: (res, err) => {
          if (!err) {
            return setAllLocally(res.value)
          }
          // TODO handle error
        },
        url: baseEndpoint + queryString
      })
    }
  }

  if (commands['Collection POST']) {
    const endpointCfg = commands['Collection POST']
    actions.create = (newEntity) => {
      return fetchAction({
        next: (res, err) => {
          if (err) {
            // TODO handle error
          } else {
            const createdEntity = res.value
            const actions = [
              addLocally(createdEntity)
            ]
            doRedirectIfNecessary({
              actions,
              endpointCfg,
              defaultStrategy: 'toEntity',
              redirectArgs: createdEntity
            })
            return actions
          }
        },
        options: {
          body: newEntity,
          method: 'POST'
        },
        url: baseEndpoint
      })
    }
  }

  if (commands['DELETE']) {
    const endpointCfg = commands['DELETE']
    actions.delete = (entity) => {
      return fetchAction({
        next: (res, err) => {
          if (err) {
            // TODO handle error
          } else {
            const actions = [
              deleteLocally(entity._id)
            ]
            doRedirectIfNecessary({
              actions,
              endpointCfg,
              defaultStrategy: 'toParent',
              redirectArgs: entity
            })
            return actions
          }
        },
        options: {
          method: 'DELETE'
        },
        url: `${baseEndpoint}/${entity._id}`
      })
    }
  }

  if (commands['GET']) {
    actions.loadOne = (id) => fetchAction({
      next: (res, err) => {
        if (!err) {
          return setLocally(res.value)
        }
        // TODO handle error
      },
      url: `${baseEndpoint}/${id}`
    })
  }

  if (commands['PUT']) {
    const endpointCfg = commands['PUT']
    actions.update = (entity) => fetchAction({
      next: (res, err) => {
        if (err) {
          // TODO handle error
        } else {
          const updatedEntity = res.value
          const actions = [
            setLocally(updatedEntity)
          ]
          doRedirectIfNecessary({
            actions,
            endpointCfg,
            defaultStrategy: 'toEntity',
            redirectArgs: updatedEntity
          })
          return actions
        }
      },
      options: {
        body: entity,
        method: 'PUT'
      },
      url: `${baseEndpoint}/${entity._id}`
    })
  }

  return actions
}

export default makeGenericModelActions
