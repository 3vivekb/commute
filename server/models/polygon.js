const Schema = require('mongoose').Schema

const userPlugin = require('./plugins/user')

const schema = new Schema({
  geometry: {
    required: true,
    type: Schema.Types.Mixed
  },
  mode: {
    required: true,
    type: String
  },
  properties: {
    required: true,
    type: Schema.Types.Mixed
  },
  siteId: {
    required: true,
    type: Schema.Types.ObjectId
  }
})

schema.plugin(userPlugin)

module.exports = schema
