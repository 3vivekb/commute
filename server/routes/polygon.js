const models = require('../models')
const makeRestEndpoints = require('../utils/restEndpoints')

module.exports = function makeRoutes (app, jwt) {
  makeRestEndpoints(app, jwt,
    {
      commands: {
        'Collection DELETE': {},
        'Collection GET': {}
      },
      model: models.Polygon,
      name: 'polygon'
    }
  )
}
