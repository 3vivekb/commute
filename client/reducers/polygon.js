import {makeGenericReducerHandlers} from '../utils/reducers'

export const reducers = makeGenericReducerHandlers({
  handlers: ['set many', 'delete many'],
  name: {
    singular: 'polygon',
    plural: 'polygons'
  }
})

export const initialState = {}
