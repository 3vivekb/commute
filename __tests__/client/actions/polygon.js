/* global describe */

import {makeGenericModelActionsTests} from '../../test-utils/actions'

import * as polygon from '../../../client/actions/polygon'

describe('actions > polygon', () => {
  makeGenericModelActionsTests({
    actions: polygon,
    commands: {
      'Collection GET': {},
      'Collection DELETE': {
        args: { siteId: 'site-id' }
      }
    },
    pluralName: 'polygons',
    singularName: 'polygon'
  })
})
