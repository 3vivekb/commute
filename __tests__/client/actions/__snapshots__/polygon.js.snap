exports[`actions > polygon generic model actions deleteMany should work 1`] = `
Object {
  "siteId": "site-id",
}
`;
